### PR TITLE
#153: Provide `Tpetra` version of `CDR_Model`

### LINKED ISSUE #119: Tempus: Provide `Tpetra` version of `Tempus_HHTAlphaTest`

### LINKED ISSUE #153: Tempus:  Provide alternative to Epetra usage in test/TestModels/CDR_Model
This introduces an epetra dependency in BackwardEuler and BDF tests. Provide tpetra alternative.

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #119 Fixes #153

### CODE REVIEW

**Review by thearusable — DISMISSED** [review]

File: packages/tempus/test/BackwardEuler/Tempus_BackwardEulerTest.cpp [hunk 1]
```diff
@@ -17,11 +18,13 @@
...
+#include "../TestModels/CDR_Model_Tpetra.hpp"
 #include "../TestModels/VanDerPolModel.hpp"
 #include "../TestUtils/Tempus_ConvergenceTestUtils.hpp"
 
 #include "Stratimikos_DefaultLinearSolverBuilder.hpp"
 #include "Thyra_LinearOpWithSolveFactoryHelpers.hpp"
+#include "Tpetra_Core.hpp"
 
 #ifdef Tempus_ENABLE_MPI
 #include "Epetra_MpiComm.h"
```
Comment (cwschilly) [inline packages/tempus/test/BackwardEuler/Tempus_BackwardEulerTest.cpp:30]: Do you still need this include?

File: packages/tempus/test/TestModels/CDR_Model_Functors.hpp [hunk 2]
```diff
@@ -0,0 +1,587 @@
+#ifndef TEMPUS_CDR_MODEL_FUNCTORS_HPP
+#define TEMPUS_CDR_MODEL_FUNCTORS_HPP
+
+#include "Kokkos_Core.hpp"
+#include "Tpetra_Details_OrdinalTraits.hpp"
+
+template<class TpetraVectorType>
+struct MeshFillFunctor
+{
+  typedef typename TpetraVectorType::impl_scalar_type Scalar;
```
Comment (cwschilly) [inline packages/tempus/test/TestModels/CDR_Model_Functors.hpp]: Could put `using Scalar = typename TpetraVectorType::impl_scalar_type;` instead (this applies to the whole file--`typedef` vs. `using`)

File: packages/tempus/test/TestModels/CDR_Model_impl.hpp [hunk 3]
```diff
@@ -355,6 +355,7 @@ void CDR_Model<Scalar>::evalModelImpl(
     if (nonnull(W_prec_out)) {
       RCP<Epetra_Operator> M_epetra = Thyra::get_Epetra_Operator(*(W_prec_out->getNonconstRightPrecOp()));
       M_inv = rcp_dynamic_cast<Epetra_CrsMatrix>(M_epetra);
+      M_inv->Print(std::cout);
```
Comment (cwschilly) [inline packages/tempus/test/TestModels/CDR_Model_impl.hpp]: I assume these were meant for debugging--just make sure to remove before putting up to Trilinos

**Comment by JacobDomagala** [comment]
Thanks @cwschilly for the comments. This PR is still in development, I've pushed the changes because I'm working on two separate machines and I want to keep the code in sync, it's not yet ready.

File: packages/tempus/test/BackwardEuler/Tempus_BackwardEulerTest.cpp (hunk 1 above)
Comment (JacobDomagala) [reply packages/tempus/test/BackwardEuler/Tempus_BackwardEulerTest.cpp:30]: Yes, this file runs both Epetra(old) and Tpetra(new) versions of CDR_Model.
Unless the SDR will want to get rid of epetra version, I'm keeping the old one for backward compability

**Comment by JacobDomagala** [comment]
@stmcgovern @cwschilly Ok I think it's ready to be reviewed

**Review by cwschilly — APPROVED** [review]
LGTM

**Comment by JacobDomagala** [comment]
I'll have to rebase the commits, one of them is not signed

**Review by stmcgovern — APPROVED** [review]

**Pull request merged**